### PR TITLE
feat(cdc): amortize snapshot read to multiple barriers

### PR DESCRIPTION
**Pull request by StrikeW**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

We ensure to poll the snapshot stream at least once in each epoch previously, which may cause barrier pile up in some scenario (https://github.com/risingwavelabs/risingwave/issues/15812)

This PR implement the [RFC](https://www.notion.so/risingwave-labs/RFC-CDC-Backfill-V2-0e93795adb0340bba1ce130918ba7138) to allow reconstruct a new snapshot stream in each `snapshot_interval` barriers, trying to mitigate the barrier pile up problem.

- Add `snapshot_interval` to configure the barrier interval to start a new snapshot
- Add `snapshot_read_full_table` to read all historical data in batches to simplify main logic
- Move the consumption of buffered events outside the backfill loop

follow-up: https://github.com/risingwavelabs/risingwave/pull/16426

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by fuyufjh** [comment]
Can you describe the background/motivation?

File: src/stream/src/executor/backfill/cdc/cdc_backfill.rs [hunk 1]
```diff
@@ -69,6 +71,11 @@ pub struct CdcBackfillExecutor<S: StateStore> {
     rate_limit_rps: Option<u32>,
 
     disable_backfill: bool,
+
+    // TODO: make these options configurable
```
Comment (StrikeW) [inline src/stream/src/executor/backfill/cdc/cdc_backfill.rs:75]: Will be done in https://github.com/risingwavelabs/risingwave/pull/16426

File: src/stream/src/from_proto/stream_cdc_scan.rs [hunk 2]
```diff
@@ -98,6 +98,8 @@ impl ExecutorBuilder for StreamCdcScanExecutorBuilder {
             state_table,
             node.rate_limit,
             disable_backfill,
+            1,
```
Comment (StrikeW) [inline src/stream/src/from_proto/stream_cdc_scan.rs:101]: `snapshot_interval` will default to `1`, and we allow user to config different value in WITH clause https://github.com/risingwavelabs/risingwave/pull/16426

**Review by StrikeW — COMMENTED** [review]
@hzxa21 @fuyufjh PTAL, will evaluate it these days.

File: src/stream/src/executor/backfill/cdc/upstream_table/snapshot.rs [hunk 3]
```diff
@@ -128,40 +113,147 @@ impl UpstreamTableRead for UpstreamTableReader<ExternalStorageTable> {
...
+                    continue;
+                } else {
+                    // Apply rate limit, see `risingwave_stream::executor::source::apply_rate_limit` for more.
+                    // May be should be refactored to a common function later.
+                    let limiter = limiter.as_ref().unwrap();
+                    let limit = read_args.rate_limit_rps.unwrap() as usize;
+
+                    // Because we produce chunks with limited-sized data chunk builder and all rows
+                    // are `Insert`s, the chunk size should never exceed the limit.
+                    assert!(chunk_size <= limit);
```
Comment (fuyufjh) [inline src/stream/src/executor/backfill/cdc/upstream_table/snapshot.rs:160]: If the last event is `Update` i.e. (an `U-` row followed by an `U+` row), the chunk_size could be `limit + 1`

https://github.com/risingwavelabs/risingwave/blob/75ebd1ea43b211b16321b444b6d47c0ed039a539/src/common/src/array/stream_chunk.rs#L181-L182

Comment (StrikeW) [reply src/stream/src/executor/backfill/cdc/upstream_table/snapshot.rs:160]: This part is a merge with main. And for the snapshot read in cdc backfill, stream chunk only contain `Insert` event, there is no `Update`.

**Review by fuyufjh — APPROVED** [review]
LGTM

**Review by hzxa21 — APPROVED** [review]
Generally LGTM

**Pull request merged**